crates/sui-core/src/authority/authority_store.rs

```diff
@@ -62,52 +69,95 @@ pub struct SuiDataStore<const ALL_OBJ_VER: bool, const USE_LOCKS: bool, S> {
     /// digest of the certificate that lead to the creation of this version of the object.
     ///
     /// When an object is deleted we include an entry into this table for its next version and
     /// a digest of ObjectDigest::deleted(), along with a link to the transaction that deleted it.
+    ///
+    /// Consistency requirements:
+    /// - Read-after-write with respect to Validator::confirmation_transaction
     parent_sync: DBMap<ObjectRef, TransactionDigest>,
 
     /// A map between the transaction digest of a certificate that was successfully processed
     /// (ie in `certificates`) and the effects its execution has on the authority state. This
     /// structure is used to ensure we do not double process a certificate, and that we can return
     /// the same response for any call after the first (ie. make certificate processing idempotent).
+    ///
+    /// Consistency requirements:
+    /// - Same as for `certificates`
     effects: DBMap<TransactionDigest, TransactionEffectsEnvelope<S>>,
 
+    // The following three tables (sequenced, schedule, last_consensus_index) are critical for
+    // sequencing shared object txes correctly, and must be updated atomically.
+
     /// Hold the lock for shared objects. These locks are written by a single task: upon receiving a valid
     /// certified transaction from consensus, the authority assigns a lock to each shared objects of the
-    /// transaction. Note that all authorities are guaranteed to assign the same lock to these objects.
+    /// transaction. Note that all authorities are guaranteed to assign the same lock to these objects,
+    /// because shared object certs are received from narwhal in the same order by all authorities,
+    /// and are processed serially in a single task.
+    ///
+    /// sequenced contains a map form (TransactionDigest, ObjectID) to the object version number
+    /// that will be created when that Transaction is sequenced.
+    /// schedule contains the next sequence number to assign for each object id, which
+    /// allows us to determine the proper sequence number to write into sequenced when receive
+    /// another shared object tx operating on the given object.
     /// TODO: These two maps should be merged into a single one (no reason to have two).
     sequenced: DBMap<(TransactionDigest, ObjectID), SequenceNumber>,
     schedule: DBMap<ObjectID, SequenceNumber>,
 
+    /// The following table is used to store a single value (the corresponding key is a constant). The value
+    /// represents the index of the latest consensus message this authority sequenced. This field is written
+    /// by a single process acting as consensus (light) client. It is used to ensure the authority processes
+    /// every message output by consensus (and in the right order).
+    last_consensus_index: DBMap<u64, ExecutionIndices>,
+
     // Tables used for authority batch structure
     /// A sequence on all executed certificates and effects.
```

Comment:
The importance of this structure, and the batch structure in terms of needed to contain all data, in causal order, and without duplicates becomes more obvious when looking at the checkpointing mechanism (and to a lesser degree only the follower mechanism).